File: ChallengesDev/C++/010-matrix-chain-multiplication/solucion.cpp
#include <vector>
#include <climits>
#include <algorithm>

/**
 * Solution to the Matrix Chain Multiplication problem using DP.
 */
class MatrixChainOptimizer {
public:
    int solve(const std::vector<int>& p) {
        int n = p.size() - 1; // Number of matrices
        std::vector<std::vector<int>> m(n + 1, std::vector<int>(n + 1, 0));

        // L is the length of the matrix chain
        for (int L = 2; L <= n; L++) {
            for (int i = 1; i <= n - L + 1; i++) {
                int j = i + L - 1;
                m[i][j] = INT_MAX;
                for (int k = i; k <= j - 1; k++) {
                    int q = m[i][k] + m[k + 1][j] + p[i - 1] * p[k] * p[j];
                    if (q < m[i][j]) {
                        m[i][j] = q;
                    }
                }
            }
        }

        return m[1][n];
    }
};

/**
 * Complexity Analysis:
 * Time: O(n^3) where n is the number of matrices.
 * Space: O(n^2) for the DP table.
 */
 pieces of code.

File: ChallengesDev/Java/010-sparse-matrix-multiplication/Solucion.java
/**
 * Solution to Challenge 010: Sparse Matrix Multiplication.
 * Optimized implementation to ignore zeros.
 */
public class Solucion {
    public int[][] multiply(int[][] mat1, int[][] mat2) {
        int m = mat1.length;
        int k = mat1[0].length;
        int n = mat2[0].length;
        
        int[][] res = new int[m][n];
        
        for (int i = 0; i < m; i++) {
            for (int r = 0; r < k; r++) {
                // Optimization: Skip if the element in mat1 is 0
                if (mat1[i][r] != 0) {
                    for (int j = 0; j < n; j++) {
                        if (mat2[r][j] != 0) {
                            res[i][j] += mat1[i][r] * mat2[r][j];
                        }
                    }
                }
            }
        }
        
        return res;
    }
}

/**
 * Complexity Analysis:
 * Time: O(m * k * n) in the worst case, but O(m * k * L) where L is the
 * average number of non-zero elements per row in mat2.
 * Space: O(m * n) to store the result.
 */
